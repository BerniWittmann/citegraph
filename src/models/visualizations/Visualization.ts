import Transformer from '@/transformers/Transformer'

export interface TimePeriod {
  from: number
  to: number
}

export interface VisualizationBaseFields {
  id?: string;
  progress?: number;
  name: string;
  timePeriods?: Array<TimePeriod>
}

export interface VisualizationBaseFieldsDTO {
  id?: string;
  progress?: number;
  name: string;
  key: string;
  // eslint-disable-next-line camelcase
  time_periods?: Array<TimePeriod>
}

export default abstract class Visualization implements VisualizationBaseFields {
  id?: string;
  progress?: number;
  name: string;
  timePeriods: Array<TimePeriod>
  abstract parameters: Object;
  abstract data?: Object;
  abstract readonly key: string;
  static readonly key: string;
  static readonly title: string;
  static readonly imageUrl: string;
  static readonly description: string;
  static readonly longDescription: string;
  static readonly transformer: typeof Transformer;

  protected constructor ({ id, progress, name, timePeriods }: VisualizationBaseFields) {
    this.id = id
    this.progress = progress
    this.name = name
    this.timePeriods = timePeriods || []
  }
}
