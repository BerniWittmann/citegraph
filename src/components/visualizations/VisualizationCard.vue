<template>
  <v-card :hover="true"
          class="visualization-card"
          height="100%">
    <div class="d-flex flex-no-wrap justify-space-between">
      <v-avatar
        class="ma-3"
        size="125"
        tile
      >
        <v-img contain :src="imageUrl"/>
      </v-avatar>
      <div class="flex-grow-1 d-flex flex-column align-center">
        <v-card-title>{{ visualization.name }}</v-card-title>
        <v-card-text class="progress-information text-center">
          <span v-if="isLoading">{{ $t('visualizations.card.processing', { percent: progress }) }}</span>
          <span v-else>{{ $t('visualizations.card.ready') }}</span>
        </v-card-text>
        <v-divider style="width: 90%" class="mx-auto"></v-divider>
        <v-card-actions>
          <v-btn color="primary" text>
            {{ $t('visualizations.card.show') }}
          </v-btn>
          <v-btn color="primary" text>
            {{ $t('visualizations.card.edit') }}
          </v-btn>
        </v-card-actions>
      </div>
    </div>
    <v-progress-linear color="#F37F21" background-color="#FBBF35" background-opacity="0.5" v-show="isLoading" class="progress" :value="progress" :max="100"></v-progress-linear>
  </v-card>
</template>
<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import Visualization from '../../models/visualizations/Visualization'
import { visualizationsKeyMap } from '@/models/visualizations'

@Component
export default class VisualizationCard extends Vue {
  @Prop() readonly visualization!: Visualization

  get visualizationClass () {
    return visualizationsKeyMap[this.visualization.key]
  }

  get imageUrl (): string {
    return `${process.env.BASE_URL}${this.visualizationClass.imageUrl}`
  }

  get progress (): number {
    return this.visualization.progress ? this.visualization.progress * 100 : 0
  }

  get isLoading (): boolean {
    return !!this.visualization.progress && (this.progress < 100)
  }
}
</script>

<style scoped lang="scss">
.progress-information {
  color: #F37F21;
}
</style>
