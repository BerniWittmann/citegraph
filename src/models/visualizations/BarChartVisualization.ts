import Visualization, { VisualizationBaseFields } from './Visualization'

export interface BarChartVisualizationFields {}

export default class BarChartVisualization extends Visualization implements BarChartVisualizationFields {
  static key: string = 'barchart'
  key: string = 'barchart'
  static title: string = 'visualizations.information.bar_chart.title'
  static description: string = 'visualizations.information.bar_chart.description'
  static longDescription: string = 'visualizations.information.bar_chart.long_description'
  static imageUrl: string = 'visualization_images/bar_chart.png'

  constructor ({ id, name, progress }: VisualizationBaseFields & BarChartVisualizationFields) {
    super({ id, name, progress })
  }
}
