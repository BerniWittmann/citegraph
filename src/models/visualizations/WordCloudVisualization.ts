import Visualization, { VisualizationBaseFields } from './Visualization'

export interface WordCloudVisualizationParameters {
  pruneLeastCitedPercentage: number
}

export interface WordCloudDataPoint {
  word: string
  weight: number
}
export interface WordCloudData {
  data: Array<WordCloudDataPoint>
}

export interface WordCloudVisualizationFields {
  data?: WordCloudData
}

export default class WordCloudVisualization extends Visualization implements WordCloudVisualizationFields {
  static key: string = 'wordcloud'
  key: string = 'wordcloud'
  static title: string = 'visualizations.information.word_cloud.title'
  static description: string = 'visualizations.information.word_cloud.description'
  static longDescription: string = 'visualizations.information.word_cloud.long_description'
  static imageUrl: string = 'visualization_images/word_cloud.jpg'
  parameters: WordCloudVisualizationParameters
  data?: WordCloudData

  constructor ({ id, name, progress, data }: VisualizationBaseFields & WordCloudVisualizationFields) {
    super({ id, name, progress })
    this.parameters = {
      pruneLeastCitedPercentage: 5
    }
    this.data = data
  }
}
