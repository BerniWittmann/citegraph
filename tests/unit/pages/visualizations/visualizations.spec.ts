import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../setupPlugins'

import VisualizationsPage from '@/pages/visualizations/Visualizations.vue'
import Visualization from '@/models/visualizations/Visualization'
import BarChartVisualization from '@/models/visualizations/BarChartVisualization'
import WordCloudVisualization from '@/models/visualizations/WordCloudVisualization'
import VisualizationCard from '@/components/visualizations/VisualizationCard.vue'
import EmptySlotComponent from '../../EmptySlotComponent.vue'

describe('pages/visualizations/VisualizationsPage.vue', () => {
  const visualizationsData = [
    new WordCloudVisualization({ id: '1', name: 'First Project' }),
    new BarChartVisualization({ id: '2', name: 'Second Project' })
  ]
  function getWrapper (visualizations?: Array<Visualization>) {
    visualizations = visualizations || visualizationsData
    return shallowMount(VisualizationsPage, {
      i18n,
      mocks: {
        $store: {
          getters: {
            'visualizations/visualizations': visualizations
          }
        },
        $route: {
          params: {
            projectId: 23
          }
        }
      },
      stubs: {
        VTooltip: EmptySlotComponent
      }
    })
  }
  it('renders the page', () => {
    const wrapper = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the visualizations', () => {
    const wrapper = getWrapper()
    const cards = wrapper.findAll(VisualizationCard)
    expect(cards.length).toEqual(2)
  })

  it('has a card to add a new visualization', () => {
    const wrapper = getWrapper()
    const addCard = wrapper.find('.add-card')
    expect(addCard.exists()).toBeTruthy()
    expect(addCard.props('to')).toEqual({ name: 'project.single.visualizations.add', params: { projectId: 23 } })
  })

  it('renders a different page if no visualizations exist', () => {
    const wrapper = getWrapper([])
    expect(wrapper.html()).toMatchSnapshot()
    const addButton = wrapper.find('v-btn-stub')
    expect(addButton.exists()).toBeTruthy()
    expect(addButton.props('to')).toEqual({ name: 'project.single.visualizations.add', params: { projectId: 23 } })
  })
})
