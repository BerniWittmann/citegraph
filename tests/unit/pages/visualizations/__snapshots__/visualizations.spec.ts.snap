// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/visualizations/VisualizationsPage.vue renders a different page if no visualizations exist 1`] = `
<div class="visualizations">
  <v-container-stub tag="div">
    <v-row-stub tag="div">
      <h1 class="display-1 mt-5">visualizations.your_visualizations</h1>
    </v-row-stub>
    <v-row-stub tag="div" justify="center">
      <v-col-stub lg="3" tag="div" class="d-flex justify-center flex-column text-center">
        <empty-icon-stub class="align-self-center mb-6"></empty-icon-stub>
        <h3 class="headline">visualizations.add.description</h3>
        <p>visualizations.add.text</p>
        <v-btn-stub color="primary" tag="button" activeclass="" to="[object Object]" large="true" rounded="true" type="button">
          <v-icon-stub>mdi-plus</v-icon-stub>
          visualizations.add.title
        </v-btn-stub>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;

exports[`pages/visualizations/VisualizationsPage.vue renders the page 1`] = `
<div class="visualizations">
  <v-container-stub tag="div">
    <v-row-stub tag="div">
      <h1 class="display-1 mt-5">visualizations.your_visualizations</h1>
    </v-row-stub>
    <v-row-stub tag="div">
      <v-col-stub cols="12" md="6" lg="3" tag="div">
        <visualization-card-stub visualization="[object Object]"></visualization-card-stub>
      </v-col-stub>
      <v-col-stub cols="12" md="6" lg="3" tag="div">
        <visualization-card-stub visualization="[object Object]"></visualization-card-stub>
      </v-col-stub>
      <v-col-stub cols="12" md="6" lg="3" tag="div">
        <div class="empty-slot-component-stub" top="">
          <v-card-stub loaderheight="4" to="[object Object]" tag="div" height="100%" minheight="130" hover="true" outlined="true" class="d-flex justify-center align-center text--primary add-card">
            <v-icon-stub color="primary">mdi-plus</v-icon-stub>
          </v-card-stub> <span>visualizations.add.title</span>
        </div>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;
